### DISCUSSION TITLE
Recursive update

### DISCUSSION THREAD

**Discussion by Mulugruntz** [geldata/gel#2392] [Q&A] [answered]
Hi!

I saw https://github.com/edgedb/edgedb/discussions/1952 but it feels like what I'm looking for is different.

Let's show a simplified SDL of what I have:

```sql
type FileAttrs {
    required property size -> int64;
}

type ImportedFile {
    required property filename -> str;
    required property n_lines -> uint64;
    required single link attrs -> FileAttrs {
        constraint exclusive;
    };
}
```

and my DDL:

```sql
FOR x IN {
    json_array_unpack(<json>$data)
}
UNION (
    INSERT ImportedFile {
        filename := <str>x[0],
        n_lines := <uint64>x[1],
        attrs := (
            INSERT FileAttrs {
                size := <uint64>x[2]
            }
        )
    }
    UNLESS CONFLICT ON (.filename)
    ELSE (
        UPDATE ImportedFile
        SET {
            n_lines := <uint64>x[1],
            attrs.size := <uint64>x[2]  # HERE! How do I do that?
        }
    )
);
```
Because I'm trying to insert a bunch of `ImportedFile`s and their corresponding `FileAttrs`. 
However, if it already exists, I want to just update it.
But I don't want to have to create `FileAttrs` all the time, I'd like to update the linked one.

Another solution would be to always do a sub-`insert` inside the `update`. And then clean the `FileAttrs` that are not linked to a `ImportedFile` in a subsequent query. But that seems pretty dirty.

What is the best / correct way of handling recursive updates?

**Comment by elprans** [comment, marked as answer]
We are mulling over adding a syntax sugar for nested updates, but for now you have to use an explicit nested `UPDATE`:

```
FOR x IN {
    json_array_unpack(<json>$data)
}
UNION (
    INSERT ImportedFile {
        filename := <str>x[0],
        n_lines := <uint64>x[1],
        attrs := (
            INSERT FileAttrs {
                size := <uint64>x[2]
            }
        )
    }
    UNLESS CONFLICT ON (.filename)
    ELSE (
        UPDATE ImportedFile
        SET {
            n_lines := <uint64>x[1],
            attrs := (UPDATE .attrs SET {.size := <uint64>x[2]})
        }
    )
);
```

**Reply by Mulugruntz** [reply]
Thanks @elprans !

Just one correction and it works:
This line 
```
attrs := (UPDATE .attrs SET {.size := <uint64>x[2]})
```
is supposed to be
```
attrs := (UPDATE .attrs SET {size := <uint64>x[2]})
```
(without the `.` before `size`)

I think it would be nice to have such an example in the docs https://www.edgedb.com/docs/datamodel/links#links

There's the many-to-one example but not the one-to-one example.
What do you think?

**Reply by elprans** [reply]
>  What do you think?

Absolutely! I opened #2401.
